### ISSUE TITLE
Add bgc ensemble programmatic-api unit test

### ISSUE THREAD

**Issue by jeff1evesque** [jeff1evesque/machine-learning#3068]
We need to unit test the bgc ensemble method, and develop any backend dependency, to allow the programmatic-api, to cooperate with the desired unit test.

**Comment by jeff1evesque** [comment]
@protojas, I began writing some unit tests for your `bgc` wrapper. Perhaps you can assist me, by working in parallel, with the related #3067. These two issues, should be similar in nature. As you're debugging committing, and debugging, we may be able to help the other, if the approaches for the issues are similar enough (I think they are). Otherwise, maybe you'll find some poorly written code along the way.

**Comment by jeff1evesque** [comment]
@protojas, if you do help with #3067, remember to `git checkout -b feature-3067 milestone-09`.

**Comment by protojas** [comment]
I can't think of any reason why you couldn't just replace instances of "bgc" with "bgr" in the code you have - the two work exactly the same, the only difference being that bagc would return a class and bagr returns a real number

**Comment by jeff1evesque** [comment]
Travis does [builds](https://travis-ci.org/jeff1evesque/machine-learning/builds/290677775#L5703-L5712), on each commit pushed to the main repository:

-  current_utc = datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S")

```bash
self.premodel_data['properties']['datetime_saved'] = current_utc

self.premodel_data['properties']['uid'] = self.uid

document = {

'properties': self.premodel_data['properties'],

>               'dataset': self.dataset

}

E AttributeError: 'DataNew' object has no attribute 'dataset'

brain/session/base_data.py:122: AttributeError
```

I suppose I haven't taken the time to work through it.

**Comment by jeff1evesque** [comment]
The programmatic-api really has two ways of ingesting data:

- you embed the json data in the `POST` request itself
- you submit a `POST` request, with a reference to a url containing the `data`
  - since the dataset for the unit tests, references the `master` branch, the unit test for this breaks

**Comment by jeff1evesque** [comment]
It seems our `model_generate` session is not working with the `bgc` model type. More specifically, after successfully submitting a dataset, for the `bgc` model type, via the `data_new` session:

```sql
MariaDB [db_machine_learning]> select * from tbl_dataset_entity;
+-----------+------------------+------------+------------+-------------+---------------------+--------------+-------------------+
| id_entity | title            | collection | model_type | uid_created | datetime_created    | uid_modified | datetime_modified |
+-----------+------------------+------------+------------+-------------+---------------------+--------------+-------------------+
|         1 | sample_bgc_title | bagc-2     |          3 |           0 | 2017-10-23 01:08:50 |         NULL | NULL              |
+-----------+------------------+------------+------------+-------------+---------------------+--------------+-------------------+
1 row in set (0.07 sec)
```

We sent another `POST` request, associated for the `model_generate`:

![bgc](https://user-images.githubusercontent.com/2907085/31868543-0bf51d96-b76e-11e7-84c7-e30b6fc7e2aa.JPG)

Then, we noticed that no corresponding `bgc` model, was cached within redis:

```bash
root@trusty64:/home/vagrant# redis-cli
127.0.0.1:6379> KEYS *
(empty list or set)
```

So, we need to understand why the `model_generate` session is not storing a serialized `bgc` model, in redis, before proceeding to debug why the `model_predict` session has errors, as shown by travis [builds](https://travis-ci.org/jeff1evesque/machine-learning/builds/291281224#L5625-L5807).

**Comment by protojas** [comment]
does that mean `clf` is not getting serialized?

**Comment by protojas** [comment]
@jeff1evesque try changing:
``` 
return bagger.fit(dataset, labels)
```
to 
```
bagger.fit(dataset, labels)
return bagger
```

**Comment by protojas** [comment]
Also I think this
```
    classifiers = [
        list_model_type[3],
        list_model_type[5],
        list_model_type[6],
        list_model_type[9]
    ]
    regressors = [
        list_model_type[4],
        list_model_type[6],
        list_model_type[7],
        list_model_type[10]
    ]
```

should be changed to

```
    classifiers = [
        list_model_type[3],
        list_model_type[5],
        list_model_type[6],
        list_model_type[9]
    ]
    regressors = [
        list_model_type[4],
        list_model_type[6],
        list_model_type[7],
        list_model_type[10]
    ]
```

based on application.yaml but I also don't think this would affect the bug that you mentioned

**Comment by protojas** [comment]
Oh wait actually it should be
```
    classifiers = [
        list_model_type[2],
        list_model_type[4],
        list_model_type[6],
        list_model_type[8]
    ]
    regressors = [
        list_model_type[3],
        list_model_type[5],
        list_model_type[7],
        list_model_type[9]
    ]
```

`bagc` is list_model_type[2] which is probably why you're getting an error, cause it's not included in the list for bagger.py

**Comment by jeff1evesque** [comment]
6aeb583: I fixed the yaml reference typos, from `bagger.py`. Now, I'm going to rerun the the unit tests manually, though I probably could have done that before committing (redundant with travis ci):

```bash
root@trusty64:/home/vagrant# cd /vagrant/test
root@trusty64:/vagrant/test# ./unit-tests
```

**Comment by jeff1evesque** [comment]
Seems the adjusted changes, from above, did boost up the unit test coverage, from 56.x% to 57.x%. However, our travis [build](https://travis-ci.org/jeff1evesque/machine-learning/builds/291760962#L5654-L5657) still indicate the same errors we were receiving earlier. That is, our `clf` object doesn't seem like it is being serialized into redis.  One way to proceed with debugging, is to possible add a `print` statement.  I'm going to try to temporarily add the following `print` statements in `brain/cache/model.py`, and see if anything is printed during a manual `cd /vagrant/test && ./unit-tests`:

```python
...
    def cache(self, hash_name, key):
        '''

        This method serializes, then caches the provided model into a redis
        hash cache.

        '''

        try:
            serialized = Converter(self.model).serialize()

            print('=======================================')
            print('repr(serialized): ' + repr(serialized))
            print('=======================================')

            self.myRedis.hset(hash_name, key, serialized)
        except Exception, error:
            self.list_error.append(str(error))
            print self.list_error

    def uncache(self, hash_name, key):
        '''

        This method unserializes, then uncaches the desired model, using
        the provided key from the redis hash cache.

        '''

        uncached = self.myRedis.hget(hash_name, key)

        print('=======================================')
        print('repr(key): ' + repr(key))
        print('repr(uncached): ' + repr(uncached))
        print('repr(hash_name): ' + repr(hash_name))
        print('=======================================')

        return Converter(uncached).deserialize()
...
```

**Comment by jeff1evesque** [comment]
@protojas, since the unit tests will lock me into this branch, will you be willing to assist with #1755, or another issue. The idea for the former, is to essentially replace python's `pickle`, with a custom json serializer. This way, we don't have to use `pickle` the `clf` object, and can instead use `json.dumps`.

**Comment by protojas** [comment]
ok i'll take a look

**Comment by jeff1evesque** [comment]
The above `print` statements, only generated the following corresponding output:

```bash
----------------------------- Captured stdout call -----------------------------
=======================================
repr(key): u'bagc-2'
repr(uncached): None
repr(hash_name): 'bagc_model'
=======================================
```

So, I'll need to force the `model_generate` tests, to fail actually, in order for the corresponding `print` statements, to output useful information, whether it was able to cache the `clf` into redis:

```python
...
    def cache(self, hash_name, key):
        '''

        This method serializes, then caches the provided model into a redis
        hash cache.

        '''

        try:
            serialized = Converter(self.model).serialize()

            print('=======================================')
            print('repr(self.model): ' + repr(self.model))
            print('repr(serialized): ' + repr(serialized))
            print('=======================================')
            raise Exception('model.py, cache()')

            self.myRedis.hset(hash_name, key, serialized)
        except Exception, error:
            self.list_error.append(str(error))
            print self.list_error

    def uncache(self, hash_name, key):
        '''

        This method unserializes, then uncaches the desired model, using
        the provided key from the redis hash cache.

        '''

        uncached = self.myRedis.hget(hash_name, key)

        return Converter(uncached).deserialize()
...
```

**Comment by jeff1evesque** [comment]
The following is a corresponding traceback output from a manual unit test:

```bash
    def cache(self, hash_name, key):
        '''

            This method serializes, then caches the provided model into a redis
            hash cache.

            '''

        try:
            serialized = Converter(self.model).serialize()

            print('=======================================')
            print('repr(self.model): ' + repr(self.model))
            print('repr(hash_name): ' + repr(hash_name))
            print('repr(key): ' + repr(key))
            print('repr(serialized): ' + repr(serialized))
            print('=======================================')

            self.myRedis.hset(hash_name, key, serialized)
        except Exception, error:
            self.list_error.append(str(error))
            print self.list_error

>       raise Exception('model.py, cache()')
E       Exception: model.py, cache()

brain/cache/model.py:68: Exception
----------------------------- Captured stdout call -----------------------------
=======================================
repr(self.model): LabelEncoder()
repr(hash_name): u'bagc_labels'
repr(key): u'bagc-1'
repr(serialized): "ccopy_reg\n_reconstructor\np1\n(csklearn.preprocessing.label\nLabelEncoder\np2\nc__b                           uiltin__\nobject\np3\nNtRp4\n(dp5\nS'classes_'\np6\ncnumpy.core.multiarray\n_reconstruct\np7\n(cnumpy\n                           ndarray\np8\n(I0\ntS'b'\ntRp9\n(I1\n(I5\ntcnumpy\ndtype\np10\n(S'U14'\nI0\nI1\ntRp11\n(I3\nS'<'\nNNNI56                           \nI4\nI0\ntbI00\nS'd\\x00\\x00\\x00e\\x00\\x00\\x00p\\x00\\x00\\x00-\\x00\\x00\\x00v\\x00\\x00\\x00a\\x                           00\\x00\\x00r\\x00\\x00\\x00i\\x00\\x00\\x00a\\x00\\x00\\x00b\\x00\\x00\\x00l\\x00\\x00\\x00e\\x00\\x00                           \\x00-\\x00\\x00\\x001\\x00\\x00\\x00d\\x00\\x00\\x00e\\x00\\x00\\x00p\\x00\\x00\\x00-\\x00\\x00\\x00v\                           \x00\\x00\\x00a\\x00\\x00\\x00r\\x00\\x00\\x00i\\x00\\x00\\x00a\\x00\\x00\\x00b\\x00\\x00\\x00l\\x00\\x                           00\\x00e\\x00\\x00\\x00-\\x00\\x00\\x002\\x00\\x00\\x00d\\x00\\x00\\x00e\\x00\\x00\\x00p\\x00\\x00\\x00                           -\\x00\\x00\\x00v\\x00\\x00\\x00a\\x00\\x00\\x00r\\x00\\x00\\x00i\\x00\\x00\\x00a\\x00\\x00\\x00b\\x00\                           \x00\\x00l\\x00\\x00\\x00e\\x00\\x00\\x00-\\x00\\x00\\x003\\x00\\x00\\x00d\\x00\\x00\\x00e\\x00\\x00\\x                           00p\\x00\\x00\\x00-\\x00\\x00\\x00v\\x00\\x00\\x00a\\x00\\x00\\x00r\\x00\\x00\\x00i\\x00\\x00\\x00a\\x0                           0\\x00\\x00b\\x00\\x00\\x00l\\x00\\x00\\x00e\\x00\\x00\\x00-\\x00\\x00\\x004\\x00\\x00\\x00d\\x00\\x00\                           \x00e\\x00\\x00\\x00p\\x00\\x00\\x00-\\x00\\x00\\x00v\\x00\\x00\\x00a\\x00\\x00\\x00r\\x00\\x00\\x00i\\                           x00\\x00\\x00a\\x00\\x00\\x00b\\x00\\x00\\x00l\\x00\\x00\\x00e\\x00\\x00\\x00-\\x00\\x00\\x005\\x00\\x0                           0\\x00'\ntbsb."
=======================================
```

Seems like they should be cached just fine, assuming all things equal with python's `pickle`. But, the following is a little curious:

```bash
jeff1evesque@XXXXXXX YYYYYAA /path/to/machine-learning (feature-3068)
$ grep -r 'bagc-1' interface/
interface/static/data/json/programmatic_interface/bagc/dataset_url/bagc-data-append.json:        "collection": "bagc-1",
interface/static/data/json/programmatic_interface/bagc/dataset_url/bagc-data-append.json:        "https://raw.githubusercontent.com/jeff1evesque/machine-learning/master/interface/static/data/json/web_interface/bagc-1.json"
interface/static/data/json/programmatic_interface/bagc/dataset_url/bagc-data-new.json:        "collection": "bagc-1",
interface/static/data/json/programmatic_interface/bagc/dataset_url/bagc-data-new.json:        "https://raw.githubusercontent.com/jeff1evesque/machine-learning/master/interface/static/data/json/web_interface/bagc-1.json"
interface/static/data/json/programmatic_interface/bagc/dataset_url/bagc-model-generate.json:        "collection": "bagc-1",
interface/static/data/json/programmatic_interface/bagc/dataset_url/bagc-model-predict.json:        "collection": "bagc-1",

jeff1evesque@XXXXXXX YYYYYAA /path/to/machine-learning (feature-3068)
$ grep -r 'bagc-2' interface/
interface/static/data/json/programmatic_interface/bagc/file_upload/bagc-data-append.json:        "collection": "bagc-2",
interface/static/data/json/programmatic_interface/bagc/file_upload/bagc-data-new.json:            "collection": "bagc-2",
interface/static/data/json/programmatic_interface/bagc/file_upload/bagc-model-generate.json:        "collection": "bagc-2",
interface/static/data/json/programmatic_interface/bagc/file_upload/bagc-model-predict.json:        "collection": "bagc-2",
```

I have a hunch that the sample dataset's `property` attributes are not set correctly.

**Comment by jeff1evesque** [comment]
Never mind, the `grep` results are consistent with the current `svm`, and `svr` implementation.

**Comment by jeff1evesque** [comment]
Stupid of me, I should have tested if the `clf` is defined, from `bagger.py`:

```python
...
    print('=======================================')
    print('repr(clf): ' + repr(clf))
    print('repr(model): ' + repr(model))
    print('repr(collection_adjusted): ' + repr(collection_adjusted))
    print('=======================================')
    raise Exception('bagger.py, baggen()')

    if clf:
        Model(clf).cache(model + '_model', collection_adjusted)
        Hset().cache(
            model + '_feature_labels',
            collection,
            json.dumps(sorted_labels)
        )
...
```

**Comment by jeff1evesque** [comment]
Great, the error resides in `bagger.py`, makes a lot of sense:

```bash
...
        print('=======================================')
        print('repr(clf): ' + repr(clf))
        print('repr(model): ' + repr(model))
        print('repr(collection): ' + repr(collection))
        print('repr(collection_adjusted): ' + repr(collection_adjusted))
        print('=======================================')
>       raise Exception('bagger.py, baggen()')
E       Exception: bagger.py, baggen()

brain/session/model/bagger.py:177: Exception
----------------------------- Captured stdout call -----------------------------
=======================================
repr(clf): None
repr(model): u'bagc'
repr(collection): u'bagc-2'
repr(collection_adjusted): u'bagc-2'
=======================================
...
```

**Comment by jeff1evesque** [comment]
@protojas can you determine if there is a syntax error on my part, or the current `bagger.py` needs slight adjustment, to ensure it does not have `None`.

**Comment by protojas** [comment]
the only reason I see right now that it would return None is if `model` were not in `regressors` or `classifiers`

maybe put an assertion `model in regressors or model in classifiers` to check if that's the case?

**Comment by protojas** [comment]
@jeff1evesque I think the yaml references were fixed in `baggen` but not in `bagging`

that's probably the problem

**Comment by jeff1evesque** [comment]
Sounds good. I'll make some adjustments tonight, and do a manual `assert` to test.

**Comment by jeff1evesque** [comment]
@protojas, adjusted `bagger.py`:

```python
#!/usr/bin/python

from flask import current_app
from brain.database.dataset import Collection
from brain.cache.hset import Hset
from brain.cache.model import Model
from sklearn import preprocessing
import sklearn.ensemble
import json


def bagging(
    dataset,
    labels,
    model,
    k=10,
    be=None,
    features=1.0,
    samples=1.0
):
    '''

    This function trains the provided dataset, with the following properties:

    @dataset, the dataset of features
    @labels, the classifications of each point in D
    @be, the base estimator (which classifier type to ensemble?)
        - default is decision trees
    @k, the number of base estimators to use
    @features, how many to features to use for each estimator. 1.0 means use
        them all.
    @samples, how many samples to draw for bagging purposes

    '''

    # local variables
    bagger = None
    list_model_type = current_app.config.get('MODEL_TYPE')
    classifiers = [
        list_model_type[2],
        list_model_type[4],
        list_model_type[6],
        list_model_type[8]
    ]
    regressors = [
        list_model_type[3],
        list_model_type[5],
        list_model_type[7],
        list_model_type[9]
    ]

    # set up the ensembler
    if model in classifiers:
        bagger = sklearn.ensemble.BaggingClassifier(
            base_estimator=be,
            n_estimators=k,
            max_samples=samples,
            max_features=features,
            verbose=0
        )
    elif model in regressors:
        bagger = sklearn.ensemble.BaggingRegressor(
            base_estimator=be,
            n_estimators=k,
            max_samples=samples,
            max_features=features,
            verbose=0
        )

    print('=======================================')
    print('bagger.py, bagging()                   ')
    print('repr(bagger): ' + repr(bagger))
    print('repr(model): ' + repr(model))
    print('repr(dataset): ' + repr(dataset))
    print('repr(labels): ' + repr(labels))
    print('=======================================')

    # train bagger
    if bagger:
        return bagger.fit(dataset, labels)


def baggen(
    model,
    collection,
    payload,
    list_error,
    feat=1.0,
    samples=1.0,
    be=None,
    bnum=10
):
    '''

    This function generates the corresponding ensemble bagger.

    '''

    # local variables
    clf = None
    sortedf = None
    labels = []
    features = []
    cursor = Collection()
    sorted_labels = False
    label_encoder = preprocessing.LabelEncoder()
    list_model_type = current_app.config.get('MODEL_TYPE')
    collection_adjusted = collection.lower().replace(' ', '_')
    datasets = cursor.query(collection_adjusted, 'aggregate', payload)
    classifiers = [
        list_model_type[2],
        list_model_type[4],
        list_model_type[6],
        list_model_type[8]
    ]
    regressors = [
        list_model_type[3],
        list_model_type[5],
        list_model_type[7],
        list_model_type[9]
    ]

    for D in datasets['result']:
        for o in D['dataset']:
            d = o['independent-variables']

            for f in d:
                if model in regressors:
                    labels.append(float(o['dependent-variable']))
                    sortedf = [float(v) for k, v in sorted(f.items())]
                    features.append(sortedf)

                elif model in classifiers:
                    labels.append(o['dependent-variable'])
                    sortedf = [v for k, v in sorted(f.items())]
                    features.append(sortedf)

                if not sorted_labels:
                    sorted_labels = [k for k, v in sorted(f.items())]

    # generate model
    if model in classifiers:
        label_encoder = preprocessing.LabelEncoder()
        label_encoder.fit(labels)
        encoded_labels = label_encoder.transform(labels)

        # create and fit
        clf = bagging(
            features,
            encoded_labels,
            model,
            be=be,
            features=feat,
            samples=samples,
            k=bnum
        )

        Model(label_encoder).cache(model + '_labels', collection_adjusted)

    elif model in regressors:
        # create and fit
        clf = bagging(
            features,
            labels,
            model,
            be=be,
            features=feat,
            samples=samples,
            k=bnum
        )

        r2 = clf.score(features, labels)
        Hset().cache(
            model + '_r2',
            collection_adjusted,
            r2
        )

    print('=======================================')
    print('bagger.py, baggen()                    ')
    print('repr(clf): ' + repr(clf))
    print('repr(model): ' + repr(model))
    print('repr(collection): ' + repr(collection))
    print('repr(collection_adjusted): ' + repr(collection_adjusted))
    print('=======================================')
    raise Exception('bagger.py, baggen()')

    if clf:
        Model(clf).cache(model + '_model', collection_adjusted)
        Hset().cache(
            model + '_feature_labels',
            collection,
            json.dumps(sorted_labels)
        )

    return {'error': list_error}
```

Manual unit tests, indicate that `clf` is `None`, and most likely because `return bagger.fit(dataset, labels)` is returning `None`. Perhaps it's an incorrect implementation of `fit`.

---

**BGC:**

```bash
        print('=======================================')
        print('bagger.py, baggen()                    ')
        print('repr(clf): ' + repr(clf))
        print('repr(model): ' + repr(model))
        print('repr(collection): ' + repr(collection))
        print('repr(collection_adjusted): ' + repr(collection_adjusted))
        print('=======================================')
>       raise Exception('bagger.py, baggen()')
E       Exception: bagger.py, baggen()

brain/session/model/bagger.py:186: Exception
----------------------------- Captured stdout call -----------------------------
=======================================
bagger.py, bagging()
repr(bagger): BaggingClassifier(base_estimator=None, bootstrap=True,
         bootstrap_features=False, max_features=1.0, max_samples=1.0,
         n_estimators=10, n_jobs=1, oob_score=False, random_state=None,
         verbose=0)
repr(model): u'bagc'
repr(dataset): [[23.45, 98.01, 0.432, 325, 54.64, 0.002, 23], [22.1, 95.96, 0.743, 342, 66.67, 0.001, 32], [20.71, 99.33, 0.648, 342, 75.67, 0.001, 30], [23.27, 95.03, 0.488, 295, 55.83, 0.001, 27], [23.27, 95.03, 0.488, 295, 55.83, 0.001, 27], [19.99, 97.78, 0.638, 303, 58.88, 0.001, 29], [22.67, 101.21, 0.832, 427, 75.45, 0.002, 26], [23.45, 98.01, 0.432, 325, 54.64, 0.002, 25], [24.32, 92.22, 0.356, 235, 64.45, 0.001, 31], [22.67, 101.21, 0.832, 427, 75.45, 0.002, 24]]
repr(labels): array([0, 3, 3, 4, 4, 4, 2, 0, 1, 2])
=======================================
=======================================
bagger.py, baggen()
repr(clf): None
repr(model): u'bagc'
repr(collection): u'bagc-2'
repr(collection_adjusted): u'bagc-2'
=======================================
```

---

**BGR:**

```bash
        print('=======================================')
        print('bagger.py, baggen()                    ')
        print('repr(clf): ' + repr(clf))
        print('repr(model): ' + repr(model))
        print('repr(collection): ' + repr(collection))
        print('repr(collection_adjusted): ' + repr(collection_adjusted))
        print('=======================================')
>       raise Exception('bagger.py, baggen()')
E       Exception: bagger.py, baggen()

brain/session/model/bagger.py:186: Exception
----------------------------- Captured stdout call -----------------------------
=======================================
bagger.py, bagging()
repr(bagger): BaggingClassifier(base_estimator=None, bootstrap=True,
         bootstrap_features=False, max_features=1.0, max_samples=1.0,
         n_estimators=10, n_jobs=1, oob_score=False, random_state=None,
         verbose=0)
repr(model): u'bagc'
repr(dataset): [[23.45, 98.01, 0.432, 325, 54.64, 0.002, 25], [24.32, 92.22, 0.356, 235, 64.45, 0.001, 31], [22.67, 101.21, 0.832, 427, 75.45, 0.002, 24], [23.45, 98.01, 0.432, 325, 54.64, 0.002, 23], [22.1, 95.96, 0.743, 342, 66.67, 0.001, 32], [20.71, 99.33, 0.648, 342, 75.67, 0.001, 30], [23.27, 95.03, 0.488, 295, 55.83, 0.001, 27], [23.27, 95.03, 0.488, 295, 55.83, 0.001, 27], [19.99, 97.78, 0.638, 303, 58.88, 0.001, 29], [22.67, 101.21, 0.832, 427, 75.45, 0.002, 26], [23.45, 98.01, 0.432, 325, 54.64, 0.002, 25], [24.32, 92.22, 0.356, 235, 64.45, 0.001, 31], [22.67, 101.21, 0.832, 427, 75.45, 0.002, 24], [23.45, 98.01, 0.432, 325, 54.64, 0.002, 23], [22.1, 95.96, 0.743, 342, 66.67, 0.001, 32], [20.71, 99.33, 0.648, 342, 75.67, 0.001, 30], [23.27, 95.03, 0.488, 295, 55.83, 0.001, 27], [23.27, 95.03, 0.488, 295, 55.83, 0.001, 27], [19.99, 97.78, 0.638, 303, 58.88, 0.001, 29], [22.67, 101.21, 0.832, 427, 75.45, 0.002, 26]]
repr(labels): array([0, 1, 2, 0, 3, 3, 4, 4, 4, 2, 0, 1, 2, 0, 3, 3, 4, 4, 4, 2])
=======================================
=======================================
bagger.py, baggen()
repr(clf): None
repr(model): u'bagc'
repr(collection): u'bagc-1'
repr(collection_adjusted): u'bagc-1'
=======================================
```

**Comment by protojas** [comment]
```
>>> qq = BaggingClassifier()
>>> if qq: print "Hello";
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/sklearn/ensemble/base.py", line 139, in __len__
    return len(self.estimators_)
AttributeError: 'BaggingClassifier' object has no attribute 'estimators_'
>>> if qq not None: print "Hello";
  File "<stdin>", line 1
    if qq not None: print "Hello";
                 ^
SyntaxError: invalid syntax
>>> if qq is not None: print "Hello";
... 
Hello

```

**Comment by jeff1evesque** [comment]
Going to further adjust:

```python
>>> import sklearn.ensemble
>>> bagger = sklearn.ensemble.BaggingClassifier()
>>> if not bagger: print 'yay'
...
yay
>>> if bagger is not None: print 'yay'
...
yay
```

**Comment by jeff1evesque** [comment]
>`AttributeError: 'BaggingClassifier' object has no attribute 'estimators_'`

Maybe because nothing was supplied to `BaggingClassifier`?

**Comment by protojas** [comment]
estimators in BaggingClassifier are created when fit() is called

estimators are the individual classifiers (decision trees) in this case that are formed when training the model N times with N variations of the training dataset

**Comment by jeff1evesque** [comment]
Travis performance is not too good today, since my manual test finished (and travis hasn't even started):

```bash
=================================== FAILURES ===================================
______________________________ test_model_predict ______________________________

client = <FlaskClient <Flask 'factory'>>
live_server = <LiveServer listening at http://localhost:52172>

    def test_model_predict(client, live_server):
        '''

        This method tests the 'model_predict' session.

        Note: for debugging, the following syntax will output the corresponding
              json values, nested within 'json.loads()', to the travis ci:

              raise ValueError(res.json['result']['key1'])

        '''

        @live_server.app.route('/load-data')
        def get_endpoint():
            return url_for('name.load_data', _external=True)

        live_server.start()

        res = client.post(
            get_endpoint(),
            headers={'Content-Type': 'application/json'},
>           data=get_sample_json('bagc-model-predict.json', 'bagc')
        )

test/live_server/3_programmatic_interface/dataset_url/pytest_bagc_dataset_url.py:150:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/lib/python2.7/dist-packages/werkzeug/test.py:801: in post
    return self.open(*args, **kw)
/usr/local/lib/python2.7/dist-packages/flask/testing.py:127: in open
    follow_redirects=follow_redirects)
/usr/local/lib/python2.7/dist-packages/werkzeug/test.py:764: in open
    response = self.run_wsgi_app(environ, buffered=buffered)
/usr/local/lib/python2.7/dist-packages/werkzeug/test.py:677: in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
/usr/local/lib/python2.7/dist-packages/werkzeug/test.py:884: in run_wsgi_app
    app_rv = app(environ, start_response)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1997: in __call__
    return self.wsgi_app(environ, start_response)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1985: in wsgi_app
    response = self.handle_exception(e)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1540: in handle_exception
    reraise(exc_type, exc_value, tb)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1982: in wsgi_app
    response = self.full_dispatch_request()
/usr/local/lib/python2.7/dist-packages/flask/app.py:1614: in full_dispatch_request
    rv = self.handle_user_exception(e)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1517: in handle_user_exception
    reraise(exc_type, exc_value, tb)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1612: in full_dispatch_request
    rv = self.dispatch_request()
/usr/local/lib/python2.7/dist-packages/flask/app.py:1598: in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
interface/views.py:94: in load_data
    response = loader.load_model_predict()
brain/load_data.py:173: in load_model_predict
    my_prediction = session.predict()
brain/session/model_predict.py:68: in predict
    return predict(model_type, self.collection, self.predictors)
brain/session/predict/sv.py:50: in predict
    collection_adjusted
brain/cache/model.py:69: in uncache
    return Converter(uncached).deserialize()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <brain.converter.model.Model object at 0x7f9efa57cc90>

    def deserialize(self):
        '''

            This method deserializes the provided object.

            '''

>       return pickle.loads(self.model)
E       UnpicklingError: could not find MARK

brain/converter/model.py:55: UnpicklingError
______________________________ test_model_predict ______________________________

client = <FlaskClient <Flask 'factory'>>
live_server = <LiveServer listening at http://localhost:51695>

    def test_model_predict(client, live_server):
        '''

        This method tests the 'model_predict' session.

        Note: for debugging, the following syntax will output the corresponding
              json values, nested within 'json.loads()', to the travis ci:

              raise ValueError(res.json['result']['key1'])

        '''

        @live_server.app.route('/load-data')
        def get_endpoint():
            return url_for('name.load_data', _external=True)

        live_server.start()

        res = client.post(
            get_endpoint(),
            headers={'Content-Type': 'application/json'},
>           data=get_sample_json('bagc-model-predict.json', 'bagc')
        )

test/live_server/3_programmatic_interface/file_upload/pytest_bagc_file_upload.py:149:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/lib/python2.7/dist-packages/werkzeug/test.py:801: in post
    return self.open(*args, **kw)
/usr/local/lib/python2.7/dist-packages/flask/testing.py:127: in open
    follow_redirects=follow_redirects)
/usr/local/lib/python2.7/dist-packages/werkzeug/test.py:764: in open
    response = self.run_wsgi_app(environ, buffered=buffered)
/usr/local/lib/python2.7/dist-packages/werkzeug/test.py:677: in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
/usr/local/lib/python2.7/dist-packages/werkzeug/test.py:884: in run_wsgi_app
    app_rv = app(environ, start_response)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1997: in __call__
    return self.wsgi_app(environ, start_response)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1985: in wsgi_app
    response = self.handle_exception(e)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1540: in handle_exception
    reraise(exc_type, exc_value, tb)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1982: in wsgi_app
    response = self.full_dispatch_request()
/usr/local/lib/python2.7/dist-packages/flask/app.py:1614: in full_dispatch_request
    rv = self.handle_user_exception(e)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1517: in handle_user_exception
    reraise(exc_type, exc_value, tb)
/usr/local/lib/python2.7/dist-packages/flask/app.py:1612: in full_dispatch_request
    rv = self.dispatch_request()
/usr/local/lib/python2.7/dist-packages/flask/app.py:1598: in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
interface/views.py:94: in load_data
    response = loader.load_model_predict()
brain/load_data.py:173: in load_model_predict
    my_prediction = session.predict()
brain/session/model_predict.py:68: in predict
    return predict(model_type, self.collection, self.predictors)
brain/session/predict/sv.py:50: in predict
    collection_adjusted
brain/cache/model.py:69: in uncache
    return Converter(uncached).deserialize()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <brain.converter.model.Model object at 0x7f9efa1b2210>

    def deserialize(self):
        '''

            This method deserializes the provided object.

            '''

>       return pickle.loads(self.model)
E       UnpicklingError: could not find MARK

brain/converter/model.py:55: UnpicklingError
=============================== warnings summary ===============================
3_programmatic_interface/results/pytest_1_svm_prediction.py::test_save_prediction
  /var/machine-learning/brain/database/query.py:281: Warning: Data truncated for column 'probability' at row 1
    self.cursor.execute(statement, sql_args)
  /var/machine-learning/brain/database/query.py:281: Warning: Data truncated for column 'decision_function' at row 1
    self.cursor.execute(statement, sql_args)

3_programmatic_interface/results/pytest_2_svr_prediction.py::test_save_prediction
  /var/machine-learning/brain/database/query.py:281: Warning: Data truncated for column 'r2' at row 1
    self.cursor.execute(statement, sql_args)

-- Docs: http://doc.pytest.org/en/latest/warnings.html
============== 2 failed, 54 passed, 3 warnings in 143.62 seconds ===============
Error: unit test exited.
```

**Comment by jeff1evesque** [comment]
I think I may have an idea, which could be a problem in either case.  You've likely developed the wrappers using the latest sklearn syntax. However, I haven't had the chance to update the [`sklearn`](https://github.com/jeff1evesque/machine-learning/blob/bf2bde583ce05d34a28f4d7869fd33d243432edc/hiera/packages.yaml#L29) version, to the current latest [`0.19.1`](https://github.com/scikit-learn/scikit-learn/releases/tag/0.19.1). Would you be interested tackling #3092? When, that's complete, I can merge `milestone-0.9` into `feature-3068`, and see if the above is resolved.  Because, just from the error, it doesn't make sense to rewind to pointer back to the 0 position.

**Comment by jeff1evesque** [comment]
I need to remember to update the following:

- `/README.md`
- `/doc/unit_test/pytest.rst`
- `/doc/programmatic_interface`
- `/doc/index.rst`

**Comment by jeff1evesque** [comment]
Ah, silly, and makes sense that the mark could not be found. Seems that I haven't created `interface/static/data/json/programmatic-interface/bagc/file_upload/bagc-model-predict.json`, nor `interface/static/data/json/programmatic-interface/bagc/dataset_url/bagc-model-predict.json`.

**Comment by protojas** [comment]
Seems like decision function doesn't work for bagging classifiers if the base estimator is None. This is because Bagging Classifier's decision function takes the average of all its base estimator's decision functions. So, if the bases are None, they won't have a decision function. I put a work around to just return None when this is the case

**Comment by jeff1evesque** [comment]
Everything looks good, at least travis succeeded on the last commit. However, the current unit tests, for the `test_model_predict`, should have additional, more specific `assert` statements:

- [`pytest_bagc_dataset_url.py`](https://github.com/protojas/machine-learning/blob/e4db3eb0ad2613de1864b92aace6d5129de3874e/test/live_server/3_programmatic_interface/dataset_url/pytest_bagc_dataset_url.py#L129-L155)
- [`pytest_bagc_file_upload`](https://github.com/protojas/machine-learning/blob/e4db3eb0ad2613de1864b92aace6d5129de3874e/test/live_server/3_programmatic_interface/file_upload/pytest_bagc_file_upload.py#L128-L152)

For, example, check out other existing `test_model_predict`:

- [`pytest_svm_file_upload.py`](https://github.com/protojas/machine-learning/blob/e4db3eb0ad2613de1864b92aace6d5129de3874e/test/live_server/3_programmatic_interface/file_upload/pytest_svm_file_upload.py#L128-L195)
- [`pytest_svm_dataset_url.py`](https://github.com/protojas/machine-learning/blob/e4db3eb0ad2613de1864b92aace6d5129de3874e/test/live_server/3_programmatic_interface/dataset_url/pytest_svm_dataset_url.py#L128-L195)

**Comment by protojas** [comment]
After running a few tests, I think it's unlikely that you'll be able to get the anywhere near the same level of detail in these assertions as you have in svm.

Here is how SVC behaves when refitting using the same data:
```
>>> svc.fit(data, classes)
SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=True, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
>>> svc.predict_proba(pred)
array([[ 0.22065272,  0.32779231,  0.45155496]])
>>> svc.fit(data, classes)
SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=True, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
>>> svc.predict_proba(pred)
array([[ 0.22065272,  0.32779231,  0.45155496]])
```
Meanwhile, here is how BaggingClassifier behaves.

```
>>> bgc.fit(data, classes)
BaggingClassifier(base_estimator=None, bootstrap=True,
         bootstrap_features=False, max_features=1.0, max_samples=1.0,
         n_estimators=10, n_jobs=1, oob_score=False, random_state=None,
         verbose=0, warm_start=False)
>>> bgc.predict_proba(pred)
array([[ 0.5,  0.5,  0. ]])
>>> bgc.fit(data, classes)
BaggingClassifier(base_estimator=None, bootstrap=True,
         bootstrap_features=False, max_features=1.0, max_samples=1.0,
         n_estimators=10, n_jobs=1, oob_score=False, random_state=None,
         verbose=0, warm_start=False)
>>> bgc.predict_proba(pred)
array([[ 0.2,  0.8,  0. ]])
>>> bgc.fit(data, classes)
BaggingClassifier(base_estimator=None, bootstrap=True,
         bootstrap_features=False, max_features=1.0, max_samples=1.0,
         n_estimators=10, n_jobs=1, oob_score=False, random_state=None,
         verbose=0, warm_start=False)
>>> bgc.predict_proba(pred)
array([[ 0.2,  0.7,  0.1]])
```

Since bagging randomly transforms the training data before fitting, I do not think you'll get any reproducible results fit for unit tests.

**Comment by jeff1evesque** [comment]
Can we provide an appropriate dataset, such that the returned values, are within some margin `x`?

**Comment by protojas** [comment]
It seems unlikely that it would work consistently. Cause, for example, you could include points A,B,C,D, and D might never be even used for training because of the randomness. Because of the randomness, I think the only time you'd get consistent responses are either by A) making the margin ridiculously large or B) attempting to predict points which were in the training dataset. Both of these are trivial test cases, though.

**Comment by jeff1evesque** [comment]
Another idea, is that we can run calculations multiple times, and average the result, and ensure the corresponding `assert` statements are within some tolerable `x`. I suppose this would be an alternate, and less preferred, since it would be more computing time, for the respective tests.

**Comment by protojas** [comment]
I also don't think that would work. If you look at the examples I posted, the same dataset could spit out 0.5 three times in a row, making an average of 0.5. It could also spit out 0.2 three times in a row, making an average of 0.2.

I mean, yea, you could average them like you said but unless the margin is like 0.5 or something (which probably isn't even worth running unit tests for) there's still a chance that the unit tests fail for no reason.

**Comment by jeff1evesque** [comment]
>Cause, for example, you could include points A,B,C,D, and D might never be even used for training because of the randomness.

This is a good point. Maybe this means, to have accurate tests, we would need to create multiple ensemble models, rather than multiple predictions of the same model.

**Comment by jeff1evesque** [comment]
I think what we are left to consider, is either create multiple models, or to generalize the testing. However, since the dataset for the tests do not have to be big, the run time is not that big of an issue? Perhaps we can average out multiple tests, along with general `assert`, like checking if the result is within the domain of provided data, for classification. How would you generalize regression based, since results could be any real number?

**Comment by jeff1evesque** [comment]
I guess making multiple ensemble models, probably doesn't make sense, unless the dataset is sufficiently big enough, to produce similar predictions. This would mean larger dataset, to reduce overfitting, which would go hand in hand with more computing time.

**Comment by protojas** [comment]
I added random_state as a parameter, when the unit test calls the function it should pass an integer, but everywhere else should pass None.. I dont really know where to put that logic though

**Comment by jeff1evesque** [comment]
The unit test, builds the flask app differently then a `vagrant up` build.  Specifically, it loads configurations from `hiera\test\hiera`, then defines corresponding configurations into the flask [`factory.py`](https://github.com/jeff1evesque/machine-learning/blob/master/factory.py#L88-L122).  This allows you to reference `from flask import current_app`, then do `current_app['flask_variable']`.  So, I guess you could define `application > sklearn > bgc > blah: 'some-value` in [`application.yaml`](https://github.com/jeff1evesque/machine-learning/blob/master/hiera/test/hiera/application.yaml#L21-L37). Then your bagger will use this value. Since, vagrant has it's won `application.yaml`, and unit testing has another, the corresponding build, will be referenced appropriately, based on how it's conditionally defined in `factory.py`.

**Comment by protojas** [comment]
random_state: 332 

the "332" doesnt mean anything in itself, it just has to be a constant integer used in all test cases.

If you use a constant throughout different bagging models, it makes the output predictable, allowing for unit tests with predictable values.

So, in the test/application.yaml, random_state = <an integer> and in the non-test application.yaml, random_state = None so that it uses an actual random number generator

**Comment by protojas** [comment]
Also decision_function should be None when using a base_estimator of None

This is because bagging relies on its internal base estimators' decision functions. If the internal ones are None, trying to call bagger.decision_function() fails, so in that case I just return None since there's no way to get output for that case.

**Comment by jeff1evesque** [comment]
This issue has been resolved per https://github.com/jeff1evesque/machine-learning/pull/3101.
